src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala

```diff
@@ -0,0 +1,524 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.fastq
+
+import java.nio.file.Files
+
+import com.fulcrumgenomics.testing.{ErrorLogLevel, SamRecordSetBuilder, UnitSpec}
+import com.fulcrumgenomics.util.{Io, Metric, ReadStructure, SampleBarcodeMetric}
+import com.fulcrumgenomics.FgBioDef.{DirPath, FilePath}
+import com.fulcrumgenomics.fastq.DemuxFastqs.UnmatchedSampleId
+import com.fulcrumgenomics.fastq.FastqDemultiplexer.DemuxRecord
+import com.fulcrumgenomics.illumina.{Sample, SampleSheet}
+import org.scalatest.OptionValues
+import dagr.commons.io.PathUtil
+import htsjdk.samtools.{SAMFileHeader, SAMReadGroupRecord, SAMUtils}
+
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
+
+class DemuxFastqsTest extends UnitSpec with OptionValues with ErrorLogLevel {
+  import DemuxFastqs._
+
+  private val sampleSheetPath: FilePath = {
+    val path = makeTempFile("SampleSheet", ".csv")
+    val lines = """
+      |[Header],,,,,,,,,
+      |IEMFileVersion,4,,,,,,,,
+      |Investigator Name,Joe,,,,,,,,
+      |Experiment Name,EXPID,,,,,,,,
+      |Date,01/01/2000,,,,,,,,
+      |Workflow,GenerateFASTQ,,,,,,,,
+      |Application,FASTQ Only,,,,,,,,
+      |Assay,Assay Name,,,,,,,,
+      |Description,The Description,,,,,,,,
+      |Chemistry,Amplicon,,,,,,,,
+      |,,,,,,,,,
+      |[Reads],,,,,,,,,
+      |151,,,,,,,,,
+      |151,,,,,,,,,
+      |,,,,,,,,,
+      |[Settings],,,,,,,,,
+      |ReverseComplement,0,,,,,,,,
+      |Adapter,AGATCGGAAGAGCACACGTCTGAACTCCAGTCA,,,,,,,,
+      |AdapterRead2,AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT,,,,,,,,
+      |,,,,,,,,,
+      |[Data],,,,,,,,,
+      |Sample_ID,Sample_Name,Sample_Plate,Sample_Well,index,index2,Sample_Barcode,Sample_Project,Description
+      |20000101-EXPID-1,Sample_Name_1,,,,,AAAAAAAAGATTACAGA,Sample_Project_1,Description_1
```

Comment:
Since you use these elsewhere it might be nice to extract the barcode sequences into four vals and use `${barcode1}` syntax to insert them here.  Then you can refer to them in tests without hard-coding them in multiple places.